### DISCUSSION TITLE
[Enhancement] Support a MAUI wrapped Blazor App on Linux

### DISCUSSION THREAD

**Discussion by RChrisCoble** [dotnet/maui#1521] [Ideas] [answered]
## Summary
This is an enhancement request, or _registering interest_ in the potential support of running a MAUI application which wraps a Blazor application running on Linux.

## Intended Use Case
In the Industrial Control space, customers want access to their Human Machine Interface (HMI) visualizations on a variety of platforms.  While the traditional UC's of consuming HMI visualizations are on Windows, Mobile, and Web platforms, there is also a great deal of UC's around smaller, light weight machines that have their own custom HMI that are packaged/delivered/(literally attached) to the specific machinery.  These light weight machines typically run versions of Linux.

## Assumptions ##
As [MS Edge has been released for Linux](https://blogs.windows.com/msedgedev/2020/10/20/microsoft-edge-dev-linux/), WebView2 [also needs to be delivered to Linux as well](https://github.com/MicrosoftEdge/WebView2Feedback/issues/645).
Some amount of MAUI needs to run on Linux as the shell hosting the WebView2 control for the Blazor application.

## High Level Vision ##
Blazor is an amazing technology stack that will allow customers to create business logic and a UI layer that spans all platforms when combined with MAUI.  For .Net 6, only desktop topologies are being targeted for a MAUI wrapped Blazor application, presumably followed by iOS/Android targets in .Net 7.  Ideally Linux support can also be considered for .Net 7.

**Comment by AlbertoPa** [comment]
I would consider this a must to compete with other technologies that already allow to wrap web apps into desktop applications and truly be cross-platform (a long-standing major drawback of .Net has been the lack of a truly cross-platform UI toolkit that does not require third-party extensions).

**Comment by GeraudFabien** [comment]
It was never say they will not support linux. There is no support for .net 6 but may be for futur release. https://github.com/dotnet/aspnetcore/issues/27234#issuecomment-765604868
Also there are already solution to wait for linux https://github.com/ElectronNET/Electron.NET

**Comment by AlbertoPa** [comment]
I did not imply it was said that there won't be support. However a) .Net 7 is over a year away b) it is long overdue to have some form of .Net UI toolkit that it truly cross-platform and comes with the .Net SDK c) Electron.Net is a third-party tool, hence it does not apply to my previous comment.

**Comment by ziaulhasanhamim** [comment]
this feature is an actual need. I think Maui should at least support blazor webview on Linux because for that there is no need for many efforts. Only a wrapper window is needed and other works are already done. This will attract many developers to maui, blazor and .net as a whole because then we'll be able to run blazor apps on the web, android, ios, windows, macos and linux, how great is that! and if you have your backend server in asp.net core then you can also share some code(Request/Response objs) from that project too. how productive is that!

**Reply by shugaoye** [reply]
I agreed. Instead of supporting everything of .NET MAUI, Blazor Hybrid app can be supported first on Linux.

**Comment by CommonLoon102** [comment]
MAUI Blazor would be obviously a superb solution over Avalonia UI (https://github.com/AvaloniaUI/Avalonia) and Eto (https://github.com/picoe/Eto). There is MS Edge for Linux (https://www.microsoftedgeinsider.com/en-us/download/?platform=linux), so having the WebView2 control as well shouldn't be a problem.

FYI: https://github.com/jsuarezruiz/maui-linux

**Reply by CommonLoon102** [reply]
Related issues:
https://github.com/jsuarezruiz/maui-linux/issues/28
https://github.com/lytico/maui/issues/9

**Comment by CrosswiredStudios** [comment]
I have clients in the medical field that are using Qt currently, and its so painful to deal with. Having a MAUI blazor app run on linux would allow me to start getting .Net tech into medical industry instruments. Its a field with large potential and visibility for showing how much better it can be.

**Reply by SliemBeji-FBC** [reply]
I come from the same field and would like to see this aswell!

**Comment by charleypeng** [comment]
I think if Microsoft wants the MAUI to be successful, linux desktop is a must-have platform.

**Comment by coderbasket** [comment]
Check out more sample here for any .Net Projects including Avalonia.
[https://coderbasket.github.io/framework/Avalonia](https://coderbasket.github.io/framework/Avalonia)

**Reply by CommonLoon102** [reply]
Not everybody likes XAML.

**Reply by ziaulhasanhamim** [reply]
Facts😂😂😂

**Reply by coderbasket** [reply]
> Not everybody likes XAML.

But it is better to have many choices... I like xaml because of WinUI,
There are many beautiful designs created with xaml like these. [https://coderbasket.github.io/framework/WinUI](https://coderbasket.github.io/framework/WinUI)

I also love Blazor and HTML, CSS, Javascript... because of lightning fast. That's why Blazor collection was created.
[https://coderbasket.github.io/framework/Blazor](https://coderbasket.github.io/framework/Blazor)

**Reply by Heartnett** [reply]
> Not everybody likes XAML.

@CommonLoon102 I agree!
The only reason why I am using Blazor is because its not XAML.

**Comment by GeraudFabien** [comment]
https://medium.com/it-dead-inside/lets-learn-blazor-creating-a-desktop-app-with-electron-net-d256c956a96a

**Reply by AlbertoPa** [reply]
Electron and Avalonia are separate frameworks, so, even if good, I don't see how they are relevant in a thread asking for *native* Linux desktop support.

**Reply by GeraudFabien** [reply]
With electron or with maui the code is run the same way. The c# code is native and the Ui isn't.

**Reply by ziaulhasanhamim** [reply]
I found electron latency is very high compared to Maui Blazor. Also, it is very resource-heavy. It runs a whole asp.net core server for no reason. Electron.net is a no-go option for real apps.

**Reply by charleypeng** [reply]
i would rather use blazor> I found electron latency is very high compared to Maui Blazor. Also, it is very resource-heavy. It runs a whole asp.net core server for no reason. Electron.net is a no-go option for real apps.

**Comment by JinShil** [comment]
Since Windows CE is now dead, the industrial sector is using ARM Embedded Linux for low spec, low cost industrial panel PCs.  Raspberry Pi Compute Module-based panel PCs are one such platform. 

It would reduce much cost, fragility, and inconvenience trying to write software for multiple industrial panel PC platforms (Windows, Embedded Linux, and sometimes Android) if Maui supported Linux.

Currently resorting to WebkitGTK+Blazor and DOM solutions like the following...
https://github.com/JinShil/BlazorWebView
https://github.com/JinShil/WebWindow
... but would prefer to not have to.

**Reply by CommonLoon102** [reply]
For industrial environment, why do you write desktop applications? Isn't it much easier to deploy the application as a web application? If you need native things like USB, serial port, etc. then it is enough to have a local service running on the client which runs a HTTP server. You allow localhost in CSP on your website, so you can get the data from the attached devices. The local service has access to the physical devices, it hosts its own HTTP server locally only. On your website you can query the local HTTP server via JavaScript which then queries a REST API, or just passing the data via JavaScript interop for Blazor-Server. In can work the other way around too.

**Reply by JinShil** [reply]
Many of the industrial machines don't even have network capabilities and don't want them.  And, yes, local resources like serial ports are essential, as is I2C, SPI, and many others.  Also, these devices are low spec, so running a full web server to provide access to local resources increases boot times and causes other latencies  that frustrate operators.

**Reply by CommonLoon102** [reply]
If they are low spec, why you want a bloated framework like MAUI wrapped Blazor? An embedded web browser is really resource hungry. Blazor is not really lightweight either. Especially the WASM one. You could just go with electron, you know.
If they don't even have network you can just go with a console application.
I can recommend this:
https://github.com/gui-cs/Terminal.Gui
Or do you need to show pictures?

.NET 6's Kestrel web server is really lightweight, it doesn't even affect boot times, as it only starts when you start your application. It is not a traditional, full-blown web server. You can just serve some APIs which respond with a JSON. You should try it out.

**Reply by JinShil** [reply]
Electron performs worse than WebkitGTK+Blazor.  A web server is unnecessary, and it most certainly does effect boot times.  I have tried it out.

**Reply by CommonLoon102** [reply]
@JinShil 
By boot time I meant OS boot time, not application startup time.
If Kestrel is too slow, then node.js could be also tried.

I've seen your WebWindow repo. Have you also tried using node.js? With that, you could host a simple HTTP server, then no need to use WebKitGTK. You open any web browser (gnome web for example (AKA `epiphany-browser`)), as it is lightweight and it is using WebKit too), and navigate to the URL of your index file. This way your application could be in any JS framework, using TypeScript is also possible, and so on.
If you need data from the host, you can call .NET DLLs from node.js, for example like this:
https://github.com/microsoft/node-api-dotnet#load-and-call-net-assemblies-from-js
AOT is also possible:
https://github.com/microsoft/node-api-dotnet#optional-net-native-aot-compilation

**Comment by jsuarezruiz** [comment]
There is a community effort here https://github.com/jsuarezruiz/maui-linux/pull/62
![image](https://github.com/dotnet/maui/assets/6755973/999c5d65-a893-4857-88c9-36ff0cb8f87c)

**Reply by HuaFangYun** [reply]
How far is MAUI running on Linux? 1 year or 2 years or 3 years?

**Reply by lytico** [reply]
estimation based on https://github.com/jsuarezruiz/maui-linux/pull/66:  around 600 - 800 hours to finalize it
just a sponor is/ contributions are missing

**Comment by lytico** [comment]
lastest draft is here: https://github.com/jsuarezruiz/maui-linux/pull/66 , Blazor App on Linux included

**Comment by mattleibow** [comment, marked as answer]
Wanted to let folks on this thread know there's now something concrete to try. Over in **[dotnet/maui-labs](https://github.com/dotnet/maui-labs)** we've been building experimental platform backends, including a **GTK4 backend for Linux** and a **native macOS AppKit backend** (macOS without Mac Catalyst). Both also ship a `BlazorWebView` host — via **WebKitGTK** on Linux and **WKWebView** on macOS — so the "run your .NET business logic natively + a Blazor UI in a native WebView" scenario works on each.

**Linux (GTK4)** — [README + setup](https://github.com/dotnet/maui-labs/blob/main/platforms/Linux.Gtk4/README.md):
- [`Microsoft.Maui.Platforms.Linux.Gtk4`](https://www.nuget.org/packages/Microsoft.Maui.Platforms.Linux.Gtk4) — the backend
- [`Microsoft.Maui.Platforms.Linux.Gtk4.BlazorWebView`](https://www.nuget.org/packages/Microsoft.Maui.Platforms.Linux.Gtk4.BlazorWebView) — BlazorWebView on WebKitGTK
- [`Microsoft.Maui.Platforms.Linux.Gtk4.Essentials`](https://www.nuget.org/packages/Microsoft.Maui.Platforms.Linux.Gtk4.Essentials) — Essentials APIs
- [`Microsoft.Maui.Platforms.Linux.Gtk4.Templates`](https://www.nuget.org/packages/Microsoft.Maui.Platforms.Linux.Gtk4.Templates) — `dotnet new` templates

**macOS (AppKit)** — [README + setup](https://github.com/dotnet/maui-labs/blob/main/platforms/MacOS/README.md):
- [`Microsoft.Maui.Platforms.MacOS`](https://www.nuget.org/packages/Microsoft.Maui.Platforms.MacOS) — the backend
- [`Microsoft.Maui.Platforms.MacOS.BlazorWebView`](https://www.nuget.org/packages/Microsoft.Maui.Platforms.MacOS.BlazorWebView) — BlazorWebView on WKWebView
- [`Microsoft.Maui.Platforms.MacOS.Essentials`](https://www.nuget.org/packages/Microsoft.Maui.Platforms.MacOS.Essentials) — Essentials APIs

All published as `0.1.0-preview` — prerelease, so add `--prerelease` when installing.

**Status — please read before building on it:**

- These backends are **experimental**.
- They are **not supported** — explicitly **not** covered by the [.NET MAUI Support Policy](https://dotnet.microsoft.com/platform/support/policy/maui), with **no support commitment** of any kind (no SLAs, no security-fix guarantees, no compatibility guarantees).
- There is **no roadmap to make them officially supported** platforms. Linux and macOS are not announced as supported .NET MAUI targets, and nothing here should be read as a plan or commitment to do so.
- APIs may change or break at any time, and a backend could be discontinued.

Building these has also been a good exercise of the extensibility work tracked in [#34099](https://github.com/dotnet/maui/issues/34099) — the places MAUI can stop actively preventing third-party platforms, without the team having to own or officially support them.

Give the experimental backends a try, file issues in [maui-labs](https://github.com/dotnet/maui-labs/issues), and let us know what's working and what's missing — with the understanding above about where things stand.
